File: dashboard/src/api/stats.ts
import { client } from './client';
import type { SummaryPoint, ByDefectPoint, ByOperatorPoint, HeatmapPoint } from '@/types';

export async function getSummary(days = 7, productId?: number): Promise<SummaryPoint[]> {
  const { data } = await client.get<SummaryPoint[]>('/stats/summary', {
    params: { days, ...(productId != null ? { product_id: productId } : {}) },
  });
  return data;
}

export async function getByDefect(days = 30, productId?: number): Promise<ByDefectPoint[]> {
  const { data } = await client.get<ByDefectPoint[]>('/stats/by-defect', {
    params: { days, ...(productId != null ? { product_id: productId } : {}) },
  });
  return data;
}

export async function getByOperator(days = 30, productId?: number): Promise<ByOperatorPoint[]> {
  const { data } = await client.get<ByOperatorPoint[]>('/stats/by-operator', {
    params: { days, ...(productId != null ? { product_id: productId } : {}) },
  });
  return data;
}

export async function getHeatmap(days = 30, productId?: number): Promise<HeatmapPoint[]> {
  const { data } = await client.get<HeatmapPoint[]>('/stats/heatmap', {
    params: { days, ...(productId != null ? { product_id: productId } : {}) },
  });
  return data;
}

File: server/app/routers/stats.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.stats import ByDefectPoint, ByOperatorPoint, HeatmapPoint, SummaryPoint
from app.services import stats as svc

router = APIRouter(prefix="/stats", tags=["stats"])

_days_q = Query(7, ge=1, le=365)


@router.get("/summary", response_model=list[SummaryPoint])
def get_summary(
    days: int = _days_q,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return svc.summary(db, days)


@router.get("/by-defect", response_model=list[ByDefectPoint])
def get_by_defect(
    days: int = _days_q,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return svc.by_defect(db, days)


@router.get("/by-operator", response_model=list[ByOperatorPoint])
def get_by_operator(
    days: int = _days_q,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return svc.by_operator(db, days)


@router.get("/heatmap", response_model=list[HeatmapPoint])
def get_heatmap(
    days: int = _days_q,
    db: Session = Depends(get_db),
    _: User = Depends(get_current_user),
):
    return svc.heatmap(db, days)
